### ISSUE TITLE
Standardization question

### ISSUE THREAD

**Issue by j9t** [WICG/nav-speculation#228]
The HTML specification knows no value “speculationrules” for the `type` attribute, and [there don’t seem to be](https://github.com/whatwg/html/issues?q=is%3Aissue+speculationrules) attempts yet to have a discussion about this and get this standardized.

Am I missing something? Would it not benefit this feature if it got more scrutiny and support?

**Comment by j9t** [comment]
Related: Have you considered a conformant way of testing this feature?

For example, using something like `<script data-type=speculation>`?

This seems safer and less obtrusive for authors that work with a baseline quality standard of outputting valid markup.

(Personally and professionally, I’d have the expectation that an expert group like the one here has this in mind, and does not need reminding of this. The community needs you to pay attention to this.)

**Comment by jeremyroman** [comment]
This proposal, if standardized, would aim to be eventually merged into the WHATWG HTML living standard. Import maps recently followed this path, and are now [described in the relevant section of HTML](https://html.spec.whatwg.org/multipage/scripting.html#attr-script-type). ECMAScript modules did something similar beforehand.

I don't follow your question about conformance. The HTML specification specifies that browsers should treat any unrecognized type as an unprocessed data block, and the conformance requirement on authors to use a MIME type is intended specifically to avoid collisions with future features added to the web platform which might use other keywords. This is just such a feature.

Using a different attribute name, like `data-type` as you propose, would of course pass any existing validation checkers, but it would do so by instead polluting the namespace [reserved for authors](https://html.spec.whatwg.org/multipage/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes):

> User agents must not derive any implementation behavior from these attributes or values. Specifications intended for user agents must not define these attributes to have any meaningful values.

HTML validators which reject unknown tag names, attribute names, etc. will of course error on content which uses newly proposed ones. I don't consider that surprising or a failure of the specification development process.

@domenic to confirm my understanding here is correct.

(Due to the holiday season, further replies to this issue might be delayed by a couple weeks; apologies for the inconvenience.)

**Comment by edent** [comment]
Given that the rules should be served with a MIME type, would it also make sense for the `<script>`  to *also* be `<script type="application/speculationrules+json">` ?

At the moment, [the spec](https://html.spec.whatwg.org/multipage/scripting.html#the-script-element) for the script element says the type must be either a MIME type, `module`, or `importmap`. I'm slightly concerned that the list will grow to include several special cases.

**Comment by domenic** [comment]
No. `type=""` contains two value spaces: specified values that affect browser behavior (like `module`, `importmap`, `speculationrules`, `webbundle`), with no `/`; and user-defined text blocks, which use a slash. Since speculation rules affects browser behavior, it's not appropriate to use a value with a slash.
